File: 0046-permutations/0046-permutations.java
class Solution {
    static List<List<Integer>> answer = new ArrayList<>(); 
    private void helper(int[] nums, int index, List<Integer> ds){
        if(ds.size() == nums.length){
            answer.add(new ArrayList<>(ds));
            return;
        }
        if(index >= nums.length ) return;
        for(int num:nums){
            if(ds.contains(num)) continue;
            ds.add(num);
            helper(nums,index+1,ds);
            ds.remove(ds.size()-1);
            // helper(nums,index+1,ds);
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        answer.clear();
        helper(nums,0,new ArrayList<>());
        return answer;
    }
}

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> result; 
    set<int> st; 
    void helper(vector<int> ds,vector<int> nums){
        if(ds.size()==nums.size()){
            result.push_back(ds);
            return;
        }

        for(int i = 0 ; i< nums.size() ; i++){
            if(st.find(nums[i])==st.end()){
                ds.push_back(nums[i]);
                st.insert(nums[i]);
                helper(ds,nums);
                ds.pop_back();
                st.erase(nums[i]);
            }
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<int> ds;
        helper(ds,nums);
        return result;
    }
};